better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)

bdhirsh opened the pull request:
There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #107917
* #106404
* #107926
* #107916
* #107915
* #104482
* __->__ #107415



cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

bdhirsh committed:
better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2728,12 +2729,23 @@ def hint(s):
+            if is_traceable_wrapper_subclass(t):
+                # If our placeholder is a tensor subclass, then the "true" symints
+                # come from the subclass's inner tensors.
+                source_to_inner_t, _ = t.__tensor_flatten__()
```
Where is the logic that tests that the specs match?

ezyang replied:
Yes I'm talking about this. Not only do you need to guard on subclassness, you need to guard on the spec being the same the next time around.

ezyang commented on test/dynamo/test_subclasses.py:
```diff
@@ -128,6 +129,119 @@ def f(x):
+                self.inner_elem = inner
+
+            def __tensor_flatten__(self):
+                return {"inner_elem": self.inner_elem}, None
```
Huh, I didn't realize we were using a dict for tensor flatten. Are you sure? Doing hash string lookups is going to be less efficient than list index lookups

ezyang replied:
Oh, is the point that the key has to exactly match the attribute name? Then why do we need to return the Tensors themselves?

bdhirsh replied:
Going with the updated version of approach (1) further down

ezyang commented on torch/_guards.py:
```diff
@@ -645,6 +645,24 @@ class ChainedSource(Source):
+        return self.inner_source.guard_source()
+
+    def name(self):
+        return f"{self.inner_source.name()}.{self.field_name}"
```
This looks very suspicious

bdhirsh replied:
yeah... see my [comment](https://github.com/pytorch/pytorch/pull/107415#discussion_r1298572969) further up - this was my attempt at propagating source information about inner tensors on subclass wrappers back to shape env / dynamo, but I'm very open to other approaches if you don't like this

ezyang commented on torch/_guards.py:
```diff
@@ -645,6 +645,24 @@ class ChainedSource(Source):
+    field_name: str
+
+    def reconstruct(self, codegen):
+        return [codegen.create_load_global(self.name(), False, add=False)]
```
This looks extremely unlikely to be correct

ezyang commented on torch/_guards.py:
```diff
@@ -645,6 +645,24 @@ class ChainedSource(Source):
+# For example, in torch.testing._internal.double_tensor.DoubleTensor,
+# fields DoubleTensor.a and DoubleTensor.b will get a WrapperSubclassFieldSource
+@dataclasses.dataclass(frozen=True)
+class WrapperSubclassFieldSource(Source):
```
Maybe we should still have a wrapper specific source, but it's not this one.

I can think of two coherent designs.

1. We fiat that all tensor subclasses MUST store their tensors as plain attributes that can be accessed with getattr. `__tensor_flatten__` returns a list of attributes to poke at to get at tensors (it does NOT return the actual tensors, because remember, we can always use getattr to get at the tensors). You don't need a WrapperSubclassFieldSource, you can just use AttrSource
2. The tensor subclasses may store tensors anywhere, and `__tensor_flatten__` is the only way you can reliably extract them. Then, if we want to test on an inner tensor, we must run `__tensor_flatten__` and then index out the particular tensor you want. You need a custom source for this. You need a fairly complicated reconstruct implementation.

(1) is more efficient but you need to change the `__tensor_flatten__` calling convention.

bdhirsh replied:
This makes sense. I was effectively trying to do (1) here, but didn't realize that I could just use an existing `AttrSource`.

It sounds like the risk of (1) is that... we're relying on them implementing `__tensor_flatten__` correctly or they risk incorrect guards, and in particular you pointed out that this requires every tensor to have an actual attr (and not be e.g. stashed in a list. This seems.. probably reasonable, with documentation?)

bdhirsh replied:
I'll read through the `Source` code some more to understand how to re-use `AttrSource`, and update `__tensor_flatten__` to only return list-of-strings.

bdhirsh replied:
Any preference on what we should pass to `__tensor_unflatten__`? The default I would probably choose is that:
- our code calls __tensor_flatten__() to get a list of attributes
- our code grabs each attribute off the tensor and transforms it in some way
- our code constructs a dict of (attribute_str, updated_tensor) and passes that dictto `__tensor_unflatten__`. Maybe we could just pass the tensors as a list instead of a dict keyed on attribute, but that doesn't clearly tell the subclass which tensor is which.

ezyang replied:
I think a dict of attribute to updated tensor is still OK for unflatten

bdhirsh replied:
this is now done

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -394,12 +398,48 @@ def is_c_of_r(complex_dtype, real_dtype):
+                        # to the creation function here.
+                        # We assume that if the inner tensors of the subclass are given symbolic sizes,
+                        # their sizes will be used to construct the (symbolic) sizes of the wrapper tensor.
+                        r = transform_subclass(
```
it would be very nice if transform_subclass had a docstring

bdhirsh replied:
will add one

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -394,12 +398,48 @@ def is_c_of_r(complex_dtype, real_dtype):
+                        # their sizes will be used to construct the (symbolic) sizes of the wrapper tensor.
+                        r = transform_subclass(
+                            t,
+                            lambda inner_src, inner_t: callback(
```
Is this a botched merge? As best as I can tell you will only get one argument to the callback

bdhirsh replied:
Darn.. yep, there was also supposed to be an update to `transform_subclass()` in this PR, that passes every (key, value) from the dict returned by __tensor_flatten__ (I accidentally committed it to a higher PR in the stack - I ran some tests for sanity before pushing, but definitely messed something up). I'm going to:

(1) move the transform_subclass() changes into this PR

(2) change it so that tensor_flatten only has to return keys (the attribute names corresponding to tensors).

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -394,12 +398,48 @@ def is_c_of_r(complex_dtype, real_dtype):
+                    if is_traceable_wrapper_subclass(t):
+
+                        def empty_create(inner_t, inner_src):
+                            is_leaf = safe_is_leaf(inner_t)
```
looks dead?

bdhirsh replied:
yep (we don't need autograd meta to be correct on the inner tensors anyway)

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -630,7 +630,7 @@ static PyObject* THPVariable_make_wrapper_subclass(
       "Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False, "
       "c10::string_view? dispatch_sizes_strides_policy=None, bool dispatch_device=False, bool dispatch_layout=False)",
-      "_make_wrapper_subclass(PyObject* cls, SymIntArrayRef size, SymIntArrayRef strides, "
+      "_make_wrapper_subclass(PyObject* cls, SymIntArrayRef size, SymIntArrayRef? strides=None, "
```
tell me more?

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -694,20 +694,35 @@ static PyObject* THPVariable_make_wrapper_subclass(
+      if (sym_storage_offset.has_value()) {
+        // Replicating what TensorImpl does with storage_offset:
+        // https://github.com/pytorch/pytorch/blob/main/c10/core/TensorImpl.cpp#L1144
+        tensor_impl->set_storage_offset(sym_storage_offset->as_int_unchecked());
```
This seems bad. You don't actually know that sym storage offset is a plain int

bdhirsh replied:
Right. I copied this because this is currently how `TensorImpl` sets the storage offset, when you pass in a `SymInt` as your storage offset (originally added a year ago [here](https://github.com/pytorch/pytorch/pull/84806/files#diff-052bd9150ef8e09289ddf644b5a6830ede49207201cd41728f6d7cc6d9cead94R899)). It seems... not quite right, but no worse than how we handle symbolic storage offsets normally today? Maybe I'm missing something

ezyang replied:
In the quoted code, it is OK to use `as_int_unchecked` because it is guarded by a conditional that storage offset must be integer. You don't have that test here!

bdhirsh replied:
Good catch. In fact, I'm removing these changes - they were left over from when I originally changed this `_make_wrapper_subclass` overload to take an optional `strides` param, but are no longer necessary.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -630,7 +630,7 @@ static PyObject* THPVariable_make_wrapper_subclass(
       "Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False, "
       "c10::string_view? dispatch_sizes_strides_policy=None, bool dispatch_device=False, bool dispatch_layout=False)",
-      "_make_wrapper_subclass(PyObject* cls, SymIntArrayRef size, SymIntArrayRef strides, "
+      "_make_wrapper_subclass(PyObject* cls, SymIntArrayRef size, SymIntArrayRef? strides=None, "
```
This change looks like it could have been done on its own, with a test on its own.

bdhirsh replied:
Yep this could have been its own PR, sorry. In fact, I'm going to remove this from the PR and not try to change this schema.

One annoying thing I ran into while working on this PR was that `make_wrapper_subclass` technically has two overloads, but the first will happily specialize on any symints, while the second won't. The second schema is a bit more annoying to use (strides are required, no kwargs). It seems like eventually, we probably want to either consolidate the two overloads, or just have the first properly support dynamic shapes - what do you think? I didn't want to handle that though so far in my PR(s), so I'm just working around it.

ezyang reviewed (changes requested):
The higher order problem with this PR is the Dynamo Source strategy

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh commented on torch/_subclasses/meta_utils.py:
```diff
@@ -459,8 +499,20 @@ def is_c_of_r(complex_dtype, real_dtype):
+                            if isinstance(t, FakeTensor):
+                                return t.fake_mode
+                            if is_traceable_wrapper_subclass(t):
+                                inner_tensors, _ = t.__tensor_flatten__()
```
I don't think so? (there's `detect_fake_mode`, but that also checks other stuff like the ambient state - I just want to check the args.

I can move this as a util into fake_tensor.py though

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh commented on torch/_subclasses/meta_utils.py:
```diff
@@ -394,12 +398,48 @@ def is_c_of_r(complex_dtype, real_dtype):
+                                inner_strides,
+                                dtype=inner_t.dtype,
+                                device="meta",
+                            )
```
I tried this, and I realized that I can't due to the check [here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L423). I don't think I fully understand the purpose of this check (it's false if either we've seen the same storage before, or our strides/storage_offset are different). In the case of storage_offset, it looks like `r` is always created to have a storage offset of 0 ([code](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L400)), so we'll always go down this path if the original tensor has a non-zero storage offset. But I can't just update that check to be `r.stride() == t.stride() and r.storage_offset() == t.storage_offset()` - since this ends up adding unnecessary guards.

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

bdhirsh committed:
Update on "better support for fakeifying and dynamoing through torch_dispatch subclasses (with dynamic shapes)"

There is already some support for plumbing `__torch_dispatch__` tensor subclasses through dynamo, but this PR beefs it up a bit and adds a test. In particular:

(1) Fakeifying tensor subclasses didn't properly set autograd metadata (requires_grad, is_leaf) on the newly fakeified wrapper subclass. I don't actually have a test for this in this PR, but it's tested pretty heavily later in my aot autograd tests

(2) Fakeifying tensor subclasses didn't properly track source information for dynamic shapes on the inner tensors. I added a new `WrapperSubclassFieldSource` subclass, that represents a source coming from a tensor field on a wrapper subclass, which I use in the fakeifying logic, and again in symbolic_shapes.py to generate proper guards.

(3) `_make_wrapper_subclass()` marginally updated this code to work better with dynamic shapes. One thing that's a bit weird about `_make_wrapper_subclass`: it has two overloads, and the first explicitly does not support dynamic shapes (and the second.. does not support kwargs). I think that later we probably want to consolidate / at least make the first overload work with dynamic shapes, but I didn't want to handle that in this PR (so these smaller changes seemed like a strict improvement).




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]
